File: packages/python/tokagentos/advanced_capabilities/form/builtins.py
"""Built-in control types for the Form capability.

Standard types available out of the box:
- text, number, email, boolean, select, date, file

Built-in types provide consistent validation, sensible defaults,
LLM extraction hints, and override protection.
"""

from __future__ import annotations

import re
from collections.abc import Callable
from typing import Any

from .types import FormControl, ValidationResult

# ---------------------------------------------------------------------------
# ControlType -- lightweight registry entry
# ---------------------------------------------------------------------------


class ControlType:
    """Unified widget/type registry entry.

    Supports three patterns:
    1. Simple types (text, number, email) -- validate/parse/format
    2. Composite types -- have sub-controls via ``get_sub_controls``
    3. External types -- have ``activate`` for async processes
    """

    def __init__(
        self,
        id: str,
        *,
        builtin: bool = False,
        validate: Callable[[Any, FormControl], ValidationResult] | None = None,
        parse: Callable[[str], Any] | None = None,
        format: Callable[[Any], str] | None = None,
        extraction_prompt: str | None = None,
        get_sub_controls: Callable[..., list[FormControl]] | None = None,
        activate: Callable[..., Any] | None = None,
        deactivate: Callable[..., Any] | None = None,
    ) -> None:
        self.id = id
        self.builtin = builtin
        self.validate_fn = validate
        self.parse_fn = parse
        self.format_fn = format
        self.extraction_prompt = extraction_prompt
        self.get_sub_controls = get_sub_controls
        self.activate = activate
        self.deactivate = deactivate

    def validate(self, value: Any, control: FormControl) -> ValidationResult:
        if self.validate_fn is not None:
            return self.validate_fn(value, control)
        return ValidationResult(valid=True)

    def parse(self, value: str) -> Any:
        if self.parse_fn is not None:
            return self.parse_fn(value)
        return value

    def format_value(self, value: Any) -> str:
        if self.format_fn is not None:
            return self.format_fn(value)
        return str(value) if value is not None else ""


# ---------------------------------------------------------------------------
# Built-in type definitions
# ---------------------------------------------------------------------------

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_DATE_RE = re.compile(
    r"^\d{4}-\d{2}-\d{2}$"
    r"|^\d{1,2}/\d{1,2}/\d{2,4}$"
    r"|^\d{1,2}-\d{1,2}-\d{2,4}$"
)


def _validate_text(value: Any, control: FormControl) -> ValidationResult:
    s = str(value) if value is not None else ""
    if control.min_length is not None and len(s) < control.min_length:
        return ValidationResult(valid=False, error=f"Minimum length is {control.min_length}")
    if control.max_length is not None and len(s) > control.max_length:
        return ValidationResult(valid=False, error=f"Maximum length is {control.max_length}")
    if control.pattern and not re.search(control.pattern, s):
        return ValidationResult(valid=False, error=f"Does not match pattern {control.pattern}")
    return ValidationResult(valid=True)


def _validate_number(value: Any, control: FormControl) -> ValidationResult:
    try:
        n = float(str(value).replace(",", "").replace("$", "").strip())
    except (ValueError, TypeError):
        return ValidationResult(valid=False, error="Not a valid number")
    if control.min is not None and n < control.min:
        return ValidationResult(valid=False, error=f"Minimum value is {control.min}")
    if control.max is not None and n > control.max:
        return ValidationResult(valid=False, error=f"Maximum value is {control.max}")
    return ValidationResult(valid=True)


def _parse_number(value: str) -> float | int:
    cleaned = value.replace(",", "").replace("$", "").strip()
    n = float(cleaned)
    return int(n) if n == int(n) else n


def _validate_email(value: Any, _control: FormControl) -> ValidationResult:
    s = str(value).strip().lower()
    if not _EMAIL_RE.match(s):
        return ValidationResult(valid=False, error="Not a valid email address")
    return ValidationResult(valid=True)


def _parse_email(value: str) -> str:
    return value.strip().lower()


def _validate_boolean(value: Any, _control: FormControl) -> ValidationResult:
    s = str(value).strip().lower()
    if s in {"true", "false", "yes", "no", "1", "0", "y", "n"}:
        return ValidationResult(valid=True)
    return ValidationResult(valid=False, error="Not a valid yes/no value")


def _parse_boolean(value: str) -> bool:
    return value.strip().lower() in {"true", "yes", "1", "y"}


def _validate_select(value: Any, control: FormControl) -> ValidationResult:
    if control.options:
        valid_values = {opt.value for opt in control.options}
        valid_labels = {opt.label.lower() for opt in control.options}
        s = str(value).strip()
        if s not in valid_values and s.lower() not in valid_labels:
            return ValidationResult(
                valid=False,
                error=f"Must be one of: {', '.join(valid_values)}",
            )
    if control.enum:
        if str(value).strip() not in control.enum:
            return ValidationResult(
                valid=False,
                error=f"Must be one of: {', '.join(control.enum)}",
            )
    return ValidationResult(valid=True)


def _validate_date(value: Any, _control: FormControl) -> ValidationResult:
    s = str(value).strip()
    if not _DATE_RE.match(s):
        return ValidationResult(
            valid=False,
            error="Not a valid date format (expected YYYY-MM-DD or similar)",
        )
    return ValidationResult(valid=True)


def _validate_file(value: Any, control: FormControl) -> ValidationResult:
    if not isinstance(value, dict):
        return ValidationResult(valid=False, error="Expected file metadata object")
    if control.file:
        if control.file.max_size and isinstance(value.get("size"), (int, float)):
            if value["size"] > control.file.max_size:
                return ValidationResult(
                    valid=False,
                    error=f"File exceeds maximum size of {control.file.max_size} bytes",
                )
    return ValidationResult(valid=True)


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------

BUILTIN_TYPES: list[ControlType] = [
    ControlType(
        id="text",
        builtin=True,
        validate=_validate_text,
        extraction_prompt="a plain text string",
    ),
    ControlType(
        id="number",
        builtin=True,
        validate=_validate_number,
        parse=_parse_number,
        format=lambda v: f"{v:,}" if isinstance(v, (int, float)) else str(v),
        extraction_prompt="a numeric value (integer or decimal)",
    ),
    ControlType(
        id="email",
        builtin=True,
        validate=_validate_email,
        parse=_parse_email,
        extraction_prompt="an email address",
    ),
    ControlType(
        id="boolean",
        builtin=True,
        validate=_validate_boolean,
        parse=_parse_boolean,
        format=lambda v: "Yes" if v else "No",
        extraction_prompt="a yes/no or true/false value",
    ),
    ControlType(
        id="select",
        builtin=True,
        validate=_validate_select,
        extraction_prompt="a choice from the available options",
    ),
    ControlType(
        id="date",
        builtin=True,
        validate=_validate_date,
        extraction_prompt="a date in YYYY-MM-DD format",
    ),
    ControlType(
        id="file",
        builtin=True,
        validate=_validate_file,
        extraction_prompt="a file upload",
    ),
]

BUILTIN_TYPE_MAP: dict[str, ControlType] = {ct.id: ct for ct in BUILTIN_TYPES}


def register_builtin_types(
    register_fn: Callable[[ControlType], None],
) -> None:
    """Register all built-in types with a FormService instance."""
    for ct in BUILTIN_TYPES:
        register_fn(ct)


def get_builtin_type(type_id: str) -> ControlType | None:
    return BUILTIN_TYPE_MAP.get(type_id)


def is_builtin_type(type_id: str) -> bool:
    return type_id in BUILTIN_TYPE_MAP

File: apps/app-form/src/builtins.ts
/**
 * @module builtins
 * @description Built-in control types for the Form Plugin
 *
 * ## Overview
 *
 * This module defines the standard control types that are available out of the box:
 * - text: Plain text strings
 * - number: Numeric values (integers or decimals)
 * - email: Email addresses with format validation
 * - boolean: Yes/no, true/false values
 * - select: Choice from predefined options
 * - date: Date values in various formats
 * - file: File uploads (handled specially)
 *
 * ## Why Built-in Types
 *
 * Built-in types provide:
 * 1. **Consistent validation** across all forms - same rules everywhere
 * 2. **Sensible defaults** for common field types - less configuration
 * 3. **LLM extraction hints** optimized for each type - better extraction
 * 4. **Override protection** to prevent accidental shadowing - safety first
 *
 * ## Architecture Decision: ControlType vs TypeHandler
 *
 * We use ControlType (not the legacy TypeHandler) because:
 * - ControlType is the unified interface for all type categories
 * - ControlType supports composite types (subcontrols)
 * - ControlType supports external types (activate/confirm)
 * - TypeHandler is legacy and maintained only for backwards compatibility
 *
 * ## Why These Specific Types
 *
 * | Type | Why Built-in |
 * |------|--------------|
 * | text | Most common field type, catch-all for strings |
 * | number | Second most common, needs special parsing (commas, $) |
 * | email | Critical for communication, has clear format rules |
 * | boolean | Binary choice, many natural language forms (yes/no/true/false) |
 * | select | Constrained choice, validation against options |
 * | date | Complex parsing (many formats), needs normalization |
 * | file | Special handling needed (size, type, storage) |
 *
 * ## Extending
 *
 * Plugins can register custom types via FormService.registerControlType().
 * Built-in types can be overridden with { allowOverride: true } option,
 * but this will log a warning.
 *
 * ## Usage
 *
 * Built-in types are automatically registered when FormService starts.
 * You don't need to call registerBuiltinTypes() manually.
 *
 * @example
 * ```typescript
 * // Check if a type is built-in before overriding
 * if (isBuiltinType('email')) {
 *   console.log('Warning: overriding built-in type');
 *   formService.registerControlType(myEmailType, { allowOverride: true });
 * }
 * ```
 */

import type { JsonValue } from "@elizaos/core";
import type { ControlType, FormControl, ValidationResult } from "./types";

// ============================================================================
// VALIDATION HELPERS
// ============================================================================

/**
 * Email regex pattern
 * WHY this pattern:
 * - Balances strictness with practicality
 * - Catches obvious errors (missing @, missing domain)
 * - Allows international characters in local part
 */
const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

/**
 * ISO date regex pattern
 * WHY this pattern:
 * - Matches YYYY-MM-DD format
 * - Common standard for data exchange
 * - LLM can normalize other formats to this
 */
const ISO_DATE_REGEX = /^\d{4}-\d{2}-\d{2}$/;

// ============================================================================
// BUILT-IN CONTROL TYPES
// ============================================================================

/**
 * Text control type
 *
 * The default type for string input. Accepts any string value.
 * Validation is typically done via pattern/minLength/maxLength on the control.
 *
 * WHY text is the default:
 * - Most form fields are ultimately strings
 * - Pattern matching handles most custom validation needs
 * - Length limits catch data quality issues
 * - Unknown types fall back to text safely
 *
 * WHY validate inside the ControlType:
 * - Centralizes string validation logic
 * - Control-level rules (minLength, maxLength, pattern) are checked here
 * - Allows consistent error messages
 */
const textType: ControlType = {
  id: "text",
  builtin: true,

  validate: (value: JsonValue, control: FormControl): ValidationResult => {
    if (value === null || value === undefined) {
      return { valid: true }; // Empty is valid; required check is separate
    }

    const str = String(value);

    // Check minLength if specified
    if (control.minLength !== undefined && str.length < control.minLength) {
      return {
        valid: false,
        error: `Must be at least ${control.minLength} characters`,
      };
    }

    // Check maxLength if specified
    if (control.maxLength !== undefined && str.length > control.maxLength) {
      return {
        valid: false,
        error: `Must be at most ${control.maxLength} characters`,
      };
    }

    // Check pattern if specified
    if (control.pattern) {
      const regex = new RegExp(control.pattern);
      if (!regex.test(str)) {
        return { valid: false, error: "Invalid format" };
      }
    }

    // Check enum if specified
    if (control.enum && !control.enum.includes(str)) {
      return {
        valid: false,
        error: `Must be one of: ${control.enum.join(", ")}`,
      };
    }

    return { valid: true };
  },

  parse: (value: string): string => String(value).trim(),

  format: (value: JsonValue): string => String(value ?? ""),

  extractionPrompt: "a text string",
};

/**
 * Number control type
 *
 * Handles numeric values including integers and decimals.
 * Supports min/max validation for value range.
 *
 * WHY special number handling:
 * - Users type "1,234" or "$50" - we need to parse through formatting
 * - LLM might extract "fifty" which needs conversion
 * - Min/max validation is numeric, not string length
 *
 * WHY parse removes commas and currency symbols:
 * - International number formats use , as thousands separator
 * - Users often include currency prefix ($, €)
 * - Agent should handle natural input, not require clean numbers
 */
const numberType: ControlType = {
  id: "number",
  builtin: true,

  validate: (value: JsonValue, control: FormControl): ValidationResult => {
    if (value === null || value === undefined || value === "") {
      return { valid: true }; // Empty is valid; required check is separate
    }

    const num = typeof value === "number" ? value : parseFloat(String(value));

    if (Number.isNaN(num)) {
      return { valid: false, error: "Must be a valid number" };
    }

    // Check min if specified
    if (control.min !== undefined && num < control.min) {
      return { valid: false, error: `Must be at least ${control.min}` };
    }

    // Check max if specified
    if (control.max !== undefined && num > control.max) {
      return { valid: false, error: `Must be at most ${control.max}` };
    }

    return { valid: true };
  },

  parse: (value: string): number => {
    const cleaned = value.replace(/[,$\s]/g, ""); // Remove common formatting
    return parseFloat(cleaned);
  },

  format: (value: JsonValue): string => {
    if (value === null || value === undefined) return "";
    const num = typeof value === "number" ? value : parseFloat(String(value));
    if (Number.isNaN(num)) return String(value);
    return num.toLocaleString();
  },

  extractionPrompt: "a number (integer or decimal)",
};

/**
 * Email control type
 *
 * Validates email format using a practical regex.
 * Not RFC-compliant but catches common errors.
 *
 * WHY simple regex (not RFC 5322):
 * - RFC-compliant regex is 1000+ chars and still misses edge cases
 * - Simple pattern catches 99% of typos (missing @, missing domain)
 * - False positives are rare and harmless
 * - Real validation is done by sending confirmation email
 *
 * WHY lowercase normalization:
 * - Email local parts are technically case-sensitive (but rarely in practice)
 * - Domain is case-insensitive by standard
 * - Consistent lowercase prevents duplicate account issues
 */
const emailType: ControlType = {
  id: "email",
  builtin: true,

  validate: (value: JsonValue): ValidationResult => {
    if (value === null || value === undefined || value === "") {
      return { valid: true }; // Empty is valid; required check is separate
    }

    const str = String(value).trim().toLowerCase();

    if (!EMAIL_REGEX.test(str)) {
      return { valid: false, error: "Invalid email format" };
    }

    return { valid: true };
  },

  parse: (value: string): string => value.trim().toLowerCase(),

  format: (value: JsonValue): string => String(value ?? "").toLowerCase(),

  extractionPrompt: "an email address (e.g., user@example.com)",
};

/**
 * Boolean control type
 *
 * Handles yes/no, true/false, on/off values.
 * LLM can normalize natural language to boolean.
 *
 * WHY accept many formats:
 * - Users say "yes", "yep", "sure", "absolutely"
 * - Different cultures prefer different affirmatives
 * - LLM normalizes to one of the accepted formats
 * - We just need to recognize the normalized output
 *
 * WHY display as "Yes"/"No":
 * - More human-readable than true/false
 * - Consistent with how agent should speak
 */
const booleanType: ControlType = {
  id: "boolean",
  builtin: true,

  validate: (value: JsonValue): ValidationResult => {
    if (value === null || value === undefined) {
      return { valid: true }; // Empty is valid; required check is separate
    }

    if (typeof value === "boolean") {
      return { valid: true };
    }

    const str = String(value).toLowerCase();
    const validValues = ["true", "false", "yes", "no", "1", "0", "on", "off"];

    if (!validValues.includes(str)) {
      return { valid: false, error: "Must be yes/no or true/false" };
    }

    return { valid: true };
  },

  parse: (value: string): boolean => {
    const str = value.toLowerCase();
    return ["true", "yes", "1", "on"].includes(str);
  },

  format: (value: JsonValue): string => {
    if (value === true) return "Yes";
    if (value === false) return "No";
    return String(value ?? "");
  },

  extractionPrompt: "a yes/no or true/false value",
};

/**
 * Select control type
 *
 * Choice from predefined options. Options come from the control definition.
 * Validates that value matches one of the allowed options.
 *
 * WHY strict option matching:
 * - Select fields have defined, constrained values
 * - Invalid selections indicate extraction error, not user error
 * - Better to reject and re-ask than accept garbage
 *
 * WHY show option labels in error:
 * - User might not know the exact value needed
 * - Showing valid options helps them choose
 * - Agent can use this for more helpful prompts
 */
const selectType: ControlType = {
  id: "select",
  builtin: true,

  validate: (value: JsonValue, control: FormControl): ValidationResult => {
    if (value === null || value === undefined || value === "") {
      return { valid: true }; // Empty is valid; required check is separate
    }

    const str = String(value);

    // Check against options if defined
    if (control.options) {
      const validValues = control.options.map((o) => o.value);
      if (!validValues.includes(str)) {
        const labels = control.options.map((o) => o.label).join(", ");
        return { valid: false, error: `Must be one of: ${labels}` };
      }
    }

    // Check against enum if defined (fallback)
    if (control.enum && !control.options) {
      if (!control.enum.includes(str)) {
        return {
          valid: false,
          error: `Must be one of: ${control.enum.join(", ")}`,
        };
      }
    }

    return { valid: true };
  },

  parse: (value: string): string => value.trim(),

  format: (value: JsonValue): string => String(value ?? ""),

  // Note: extractionPrompt is typically customized per-field with option labels
  extractionPrompt: "one of the available options",
};

/**
 * Date control type
 *
 * Handles date values in ISO format (YYYY-MM-DD).
 * LLM normalizes various date formats to ISO.
 *
 * WHY ISO format as canonical:
 * - Unambiguous (no 01/02 confusion between US/EU)
 * - Sortable as strings
 * - Standard for data exchange
 *
 * WHY flexible parsing:
 * - Users say "tomorrow", "next Monday", "12/25/2024"
 * - LLM should normalize to ISO
 * - We accept anything Date() can parse as fallback
 *
 * WHY locale display:
 * - Agent should speak in user's date format
 * - toLocaleDateString() adapts automatically
 */
const dateType: ControlType = {
  id: "date",
  builtin: true,

  validate: (value: JsonValue): ValidationResult => {
    if (value === null || value === undefined || value === "") {
      return { valid: true }; // Empty is valid; required check is separate
    }

    const str = String(value);

    // Check ISO format
    if (!ISO_DATE_REGEX.test(str)) {
      return { valid: false, error: "Must be in YYYY-MM-DD format" };
    }

    // Check if it's a valid date
    const date = new Date(str);
    if (Number.isNaN(date.getTime())) {
      return { valid: false, error: "Invalid date" };
    }

    return { valid: true };
  },

  parse: (value: string): string => {
    // Try to parse and normalize to ISO
    const date = new Date(value);
    if (!Number.isNaN(date.getTime())) {
      return date.toISOString().split("T")[0];
    }
    return value.trim();
  },

  format: (value: JsonValue): string => {
    if (!value) return "";
    const date = new Date(String(value));
    if (Number.isNaN(date.getTime())) return String(value);
    return date.toLocaleDateString();
  },

  extractionPrompt: "a date (preferably in YYYY-MM-DD format)",
};

/**
 * File control type
 *
 * Placeholder for file uploads. Actual file handling is done
 * by the file upload pipeline, not the form system.
 *
 * Validation here is basic; real validation happens during upload.
 *
 * WHY separate from other types:
 * - File content isn't serializable in session state
 * - Actual upload is handled by messaging platform
 * - We only store metadata (name, size, mimeType, url)
 *
 * WHY validation is minimal:
 * - Real file validation happens at upload time
 * - By the time we see it, it's already uploaded
 * - We just validate the metadata structure
 *
 * WHY display as "N file(s)":
 * - File content isn't meaningful to show
 * - Count and names are what user cares about
 */
const fileType: ControlType = {
  id: "file",
  builtin: true,

  validate: (value: JsonValue, _control: FormControl): ValidationResult => {
    if (value === null || value === undefined) {
      return { valid: true }; // Empty is valid; required check is separate
    }

    // For file type, value is typically a FieldFile or array of FieldFile
    // Basic validation - detailed validation happens in upload pipeline
    if (typeof value === "object") {
      return { valid: true };
    }

    return { valid: false, error: "Invalid file data" };
  },

  format: (value: JsonValue): string => {
    if (!value) return "";
    if (Array.isArray(value)) {
      return `${value.length} file(s)`;
    }
    if (typeof value === "object" && value !== null && "name" in value) {
      return String((value as { name: string }).name);
    }
    return "File attached";
  },

  extractionPrompt: "a file attachment (upload required)",
};

// ============================================================================
// EXPORTS
// ============================================================================

/**
 * Array of all built-in control types.
 *
 * These are registered automatically when FormService starts.
 * The order doesn't matter; they're looked up by id.
 *
 * WHY array + map:
 * - Array allows easy iteration for registration
 * - Map provides O(1) lookup for override checks
 * - Both derived from same source ensures consistency
 */
export const BUILTIN_TYPES: ControlType[] = [
  textType,
  numberType,
  emailType,
  booleanType,
  selectType,
  dateType,
  fileType,
];

/**
 * Map of built-in types by id for quick lookup.
 *
 * WHY Map (not object):
 * - Type-safe keys
 * - O(1) lookup performance
 * - Clear has/get semantics
 */
export const BUILTIN_TYPE_MAP: Map<string, ControlType> = new Map(
  BUILTIN_TYPES.map((t) => [t.id, t])
);

/**
 * Register all built-in types with a FormService instance.
 *
 * This is called automatically during FormService.start().
 * You typically don't need to call this directly.
 *
 * WHY take a function (not FormService):
 * - Avoids circular dependency (builtins.ts ↔ service.ts)
 * - Service passes its registerControlType method
 * - Clean separation of concerns
 *
 * @param registerFn - The FormService.registerControlType method
 */
export function registerBuiltinTypes(
  registerFn: (type: ControlType, options?: { allowOverride?: boolean }) => void
): void {
  for (const type of BUILTIN_TYPES) {
    registerFn(type);
  }
}

/**
 * Get a built-in type by id.
 *
 * This is a convenience for checking if a type is built-in
 * before attempting to override it.
 *
 * WHY this exists:
 * - Plugins may want to extend a built-in type
 * - Checking before override allows informed decisions
 * - Avoids accidental shadowing
 *
 * @param id - The type id to look up
 * @returns The ControlType if found, undefined otherwise
 */
export function getBuiltinType(id: string): ControlType | undefined {
  return BUILTIN_TYPE_MAP.get(id);
}

/**
 * Check if a type id is a built-in type.
 *
 * WHY boolean convenience:
 * - Most callers just need yes/no answer
 * - Cleaner than checking getBuiltinType() !== undefined
 *
 * @param id - The type id to check
 * @returns true if the type is built-in
 */
export function isBuiltinType(id: string): boolean {
  return BUILTIN_TYPE_MAP.has(id);
}
